File: python/questions/findAnagrams.py
from typing import List
from collections import defaultdict

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        lp = len(p)
        ans = []
        l = 0
        r = 0
        dt = defaultdict(int)
        for c in p:
            dt[c] += 1
        while r < len(s):
            dt[s[r]] -= 1
            while dt[s[r]] < 0:
                dt[s[l]] += 1
                l += 1
            if r-l == lp-1:
                ans.append(l)
            r += 1
        return ans

File: cpp/questions/findAnagrams.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int ps = p.size();
        int l = 0;
        int r = 0;
        std::vector<int> ans; 
        int cc[26] = {0};
        for (const auto c: p) {
            cc[c-'a']++;
        }
        while (r < s.size()) {
            cc[s[r]-'a']--;
            while (cc[s[r]-'a'] < 0) {
                cc[s[l++]-'a']++;
            }
            if (++r-l == ps) {
                ans.push_back(l);
            }
        }
        return ans;
    }
};
